### DISCUSSION TITLE
Return only known properties

### DISCUSSION THREAD

**Discussion by fredericgboutin-yapla** [justinrainbow/json-schema#729] [Q&A]
Hi,
One use case I end-up with constantly is that, sure I validate my JSON when parsing it but the code that follows and uses the parsed document might use properties that are not defined in my JSON schema (thus not validated) because I'm trying to keep the JSON schema as short as possible. The reasoning is that I don't want to / I cannot document the entire schema but rather a subset of what I'm using so the validation won't fail on things I don't use anyway.

So, I basically need to filter my JSON document after validating it based on the defined JSON schema properties and I thought, maybe there is something in there for me in `justinrainbow/json-schema` but no luck. There are several very valuable flags / constraint but none that would "filter out not validated data".

Do you guys have any experience with this?

Thank you!

**Comment by fredericgboutin-yapla** [comment]
This is a bit of BS but ATM I think I can live with a simple filtering method done post validation and based on the schema,

```php
function filter($data, \stdClass $schema)
{
	if (!isset($schema->type)) {
		// We don't support things like https://json-schema.org/understanding-json-schema/structuring#dollarref, only explicit types
		throw new \Exception('No type defined in schema');
	}

	switch ($schema->type) {
		case 'object':
			if (!is_object($data)) {
				throw new \Exception('Unable to filter properties on non-object');
			}
			
			$filteredData = new \stdClass();

			// Filter all properties
			foreach ($schema->properties ?? [] as $propertyName => $propertyDefinition) {
				if (property_exists($data, $propertyName)) {
					$filteredData->$propertyName = filter($data->$propertyName, $propertyDefinition);
				}
			}

			return $filteredData;
		case 'array':
			if (!is_array($data)) {
				throw new \Exception('Unable to filter items on non-array');
			}

			$filteredData = [];
			
			// Filter all items 
			foreach ($data as $item) {
				$filteredData[] = filter($item, $schema->items);
			}

			return $filteredData;				
		default:
			// This includes: string, numeric types, boolean and null
			return $data;
	}
}

filter(
	json_decode($jsonDocument, false),
	json_decode($jsonSchema, false),
);
```

**Reply by DannyvdSluijs** [reply]
I've been overthinking how to answer this over the past couple of days. And let me start of by saying this is a really nice feature. Which currently isn't supported by `json-schema`.

This project is recovering from a long period of inactivity. The inactivity is most because of time-constraints from the contributors.
During the reboot of the project we've cleaned the issue list from several feature requests similar to what is being suggested here. That cleanup also made us realise we should stick to the core which is JSON schema's van validation.  Therefore I don't see this suggestion making it into this library.

However having said that I would still be very okay to promote other libraries which build on top of `json-schema` and think this would also help this library to grow and help others in finding the right extension that would fit their (edge-) case.

Probably not the answer you were hoping for but I hope it highlights my/our view of things.
Please, let us know if this suffices or if you would like to continue the discussion.

**Reply by fredericgboutin-yapla** [reply]
The problem is that, after creating this piece of code, I came to realize this is pretty much what can be accomplished without re-writing what your library is doing. Most notably, in order to add support for filtering [`$ref`](https://json-schema.org/understanding-json-schema/structuring#dollarref) elements.

I suppose that if we want to keep the library simple, we would need to implement some sort of visitor pattern or a mecanism so we can have custom code being called upon JSON document parsing so we could add filter logic in the produced output. 🤔

**Reply by DannyvdSluijs** [reply]
My only concern is that adding a system that allows to hook into json-schema is something beyond its core. 

The current workload ahead (version 6, supporting newer PHP versions en json-schema drafts) and  minimum time from contributors makes that such an addition is beyond our priorities and capacity. 

Sorry to let you down. My offer still remains if you build something awesome on top of json-schema we can find a place here to promote it.
